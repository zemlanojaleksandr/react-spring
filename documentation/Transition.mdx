---
name: Transition
route: /transition
order: 6
---

import { PropsTable } from 'docz'
import { Transition, config, animated } from '../src/targets/web'
import { RewindSpringProvider, RewindSpring } from './resources/RewindSpring.js'
import CodeMirror from './resources/Codemirror'
import DemoGrid from '../examples/components/DemoGrid'
import Demo from '../examples/components/Demo'
import examples from '../examples/components/examples'

# Transition

```jsx
import { Transition } from 'react-spring'
```

<RewindSpringProvider>
  <p>
    A <code>Transition</code> animates component lifecycles as components mount, unmount or otherwise change.
  </p>
  <p><span className="highlight">It takes a list of items of any type, and their keys</span>. The latter defaults to <code>item => item</code>, which, if your items are self-sufficient as a key is often good enough. Whenever items are added, removed, reordered or updated, it will help you to animate these changes.
  </p>
  <p>
     <span className="highlight">You provide a single function-child</span> that receives an item and additionally its transition state (enter/leave/update), followed by its index. The full signature looks like:
  </p>

  <CodeMirror value={`(item, state, index) => props => ReactNode`}/>
  
  <p style={{ marginTop: '1em' }}>
    <span className="highlight">You should always rely on the item that you receive</span> instead of reading from scope, since you could be rendering an item that has already been deleted, but is retained inside Transition until its fade-out is complete. This is especially important for routes.
  </p>
  <p>
    TLDR, items in, keys if necessary, present the item that comes out and apply the animated props you receive.
  </p>

  <div className="code-table">
    <CodeMirror value={`const items = [...]
  
<Transition
  items={items} keys={item => item.key}
  from={{ transform: 'translate3d(0,-40px,0)' }}
  enter={{ transform: 'translate3d(0,0px,0)' }}
  leave={{ transform: 'translate3d(0,-40px,0)' }}>
  {item => props =>
    <div style={props}>{item.text}</div>
  }
</Transition>`}/>
    <RewindSpring>
      {x => (
        <>
          <animated.div style={{ opacity: x.interpolate({ range: [0.75,1.0], output: [0,1] }), transform: x.interpolate({ range: [0.75,1.0], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>1</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.25,0.5], output: [0,1] }), transform: x.interpolate({ range: [0.25,0.5], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>2</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.0,0.25], output: [0,1] }), transform: x.interpolate({ range: [0.0,0.25], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>3</animated.div>
          <animated.div style={{ opacity: x.interpolate({ range: [0.5,0.75], output: [0,1] }), transform: x.interpolate({ range: [0.5,0.75], output: [-40,0], extrapolate: 'clamp' }).interpolate(x => `translate3d(0,${x}px,0)`) }}>4</animated.div>
        </>
      )}
    </RewindSpring>
  </div>

  <p>
    <span className="highlight">Transitions can also take a single item</span>, which can be anything. You can use it to toggle between components.
  </p>

  <div className="code-table">
    <CodeMirror value={`<Transition
  items={toggle}
  from={{ position: 'absolute', opacity: 0 }}
  enter={{ opacity: 1 }}
  leave={{ opacity: 0 }}>
  {toggle =>
    toggle
      ? props => <div style={props}>😄</div>
      : props => <div style={props}>🤪</div>
  }
</Transition>`}/>
    <RewindSpring>
      {x => (
        <>
          <animated.div style={{ position: 'absolute', opacity: x.interpolate({ range: [0.0,1.0], output: [0,1] }) }}>😄</animated.div>
          <animated.div style={{ position: 'absolute', opacity: x.interpolate({ range: [1.0,0.0], output: [0,1] }) }}>🤪</animated.div>
        </>
      )}
    </RewindSpring>
  </div>

  <p>
    It also comes in handy for <span className="highlight">single-component mount/unmount reveals</span>.
  </p>

  <div className="code-table">
    <CodeMirror value={`<Transition
  items={show}
  from={{ opacity: 0 }}
  enter={{ opacity: 1 }}
  leave={{ opacity: 0 }}>
  {show =>
    show && (props => <div style={props}>✌️</div>)
  }
</Transition>`}/>
    <RewindSpring>
      {x => <animated.div style={{ opacity: x }}>✌️</animated.div>}
    </RewindSpring>
  </div>

</RewindSpringProvider>

## Multistage transitions

<p>
  <span className="highlight">Transitions inherit from <a href="/keyframes">Keyframes</a></span>, so its slots (initial, from, enter, leave and update) can take the same types that Keyframe slots can take, namely: objects, functions, object-arrays. This gives you the ability to handle complex multistage, chained or scripted animations.
</p>

<br/>

<iframe
  src="https://codesandbox.io/embed/jjkj3x9o55?fontsize=13&editorsize=60&hidenavigation=1&codemirror=1"
  style={{ width: '100%', height: 500, border: 0, borderRadius: 4, overflow: 'hidden' }}
  sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin" />

## Props

<p>
  <span className="highlight">Transition accepts <i>all</i> spring properties</span>: native, from, immediate, onRest, and so on.
</p>

<PropsTable of={Transition} />

## Demos

<DemoGrid padding={0}>
  {examples.filter(data => data.tags.includes('transitions')).map(data => (
    <Demo
      key={data.name}
      {...data}
      link={`https://github.com/drcmda/react-spring/tree/master/examples/demos/${data.name}`}
      import={import('../examples/demos/' + data.name)}
    />
  ))}
</DemoGrid>