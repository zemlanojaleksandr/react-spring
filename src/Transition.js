import React from 'react'
import PropTypes from 'prop-types'
import Spring, { config as springConfig } from './Spring'

const empty = () => null

const ref = (object = {}, key) =>
  typeof object === 'function' ? object(key) : object

const get = props => {
  let { keys, children, render, items, ...rest } = props
  children = render || children || empty
  keys = typeof keys === 'function' ? items.map(keys) : keys
  if (!Array.isArray(children)) {
    children = [children]
    keys = keys ? [keys] : children.map(c => c.toString())
  }
  return { keys, children, items, ...rest }
}

export default class Transition extends React.PureComponent {
  static propTypes = {
    native: PropTypes.bool,
    config: PropTypes.object,
    from: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    enter: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    leave: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    update: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    keys: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.number])
      ),
      PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    ]),
    items: PropTypes.oneOfType([
      PropTypes.arrayOf(
        PropTypes.oneOfType([
          PropTypes.string,
          PropTypes.number,
          PropTypes.object,
        ])
      ),
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.object,
      ]),
    ]),
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.func),
      PropTypes.func,
    ]),
    render: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.func),
      PropTypes.func,
    ]),
  }

  constructor(prevProps) {
    super()
    this.springs = []
    this.state = { transitions: [], prevProps }
  }

  static getDerivedStateFromProps(props, { transitions, prevProps }) {
    const { keys, children, items, from, enter, leave, update } = get(props)
    const { keys: _keys, children: _children, items: _items } = get(prevProps)

    // Compare next keys with current keys
    let allKeys = transitions.map(t => t.key)
    let nextSet = new Set(keys)
    let currentSet = new Set(allKeys)
    let added = keys.filter(item => !currentSet.has(item))
    let deleted = allKeys.filter(item => !nextSet.has(item))
    let rest = keys.filter(item => currentSet.has(item))

    // Insert new keys into the transition collection
    added.forEach(key => {
      const i = keys.indexOf(key)
      transitions = [...transitions.slice(0, i), key, ...transitions.slice(i)]
    })

    transitions = transitions.map(transition => {
      const isTransition = typeof transition === 'object'
      const key = isTransition ? transition.key : transition
      const keyIndex = keys.indexOf(key)
      const item = items ? items[keyIndex] : key
      if (isTransition) {
        // A transition already exists
        if (deleted.find(k => k === key)) {
          // The transition was removed, re-key it and animate it out
          return {
            ...transition,
            destroyed: true,
            prevKey: key,
            key: transition.key + '_',
            to: !transition.destroyed
              ? ref(leave, _items ? _items[_keys.indexOf(key)] : key)
              : transition.to,
          }
        }
        // Transition remains untouched, update children and call hook
        return {
          ...transition,
          children: children[keyIndex] || transition.children,
          to:
            update && rest.indexOf(transition.key) !== -1
              ? ref(update, item) || transition.to
              : transition.to,
        }
      }
      // Map added key into transition
      return {
        children: children[keyIndex],
        key,
        item,
        to: ref(enter, item),
        from: ref(from, item),
      }
    })

    // Re-order list
    let ordered = keys.map(key => transitions.find(child => child.key === key))
    transitions.forEach((t, i) => {
      if (t.destroyed)
        ordered = [...ordered.slice(0, i), t, ...ordered.slice(i)]
    })

    return { transitions: ordered, prevProps: props }
  }

  getValues() {
    return undefined
  }

  render() {
    const {
      render,
      from = {},
      enter = {},
      leave = {},
      native = false,
      config = springConfig.default,
      keys,
      items,
      onFrame,
      onRest,
      ...extra
    } = this.props
    const props = { native, config, ...extra }
    return this.state.transitions.map((transition, i) => {
      const { prevKey, key, item, children, from, ...rest } = transition
      return (
        <Spring
          ref={r => r && (this.springs[key] = r)}
          key={key}
          onRest={
            rest.destroyed
              ? () =>
                  this.setState(
                    ({ transitions }) => ({
                      transitions: transitions.filter(t => t !== transition),
                    }),
                    () => delete this.springs[key]
                  )
              : onRest && (values => onRest(item, values))
          }
          onFrame={onFrame && (values => onFrame(item, values))}
          {...rest}
          {...props}
          from={rest.destroyed ? this.springs[prevKey].getValues() : from}
          render={render && children}
          children={render ? this.props.children : children}
        />
      )
    })
  }
}
