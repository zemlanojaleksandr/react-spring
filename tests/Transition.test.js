import React from 'react'
import { shallow } from 'enzyme'
import Transition from '../src/Transition'

const fakeChild = item => props => {
  ;<div>fake</div>
}

// These snapshots won't reveal much, but they at least will catch hard errors.
describe('Transition', () => {
  it('smoke/regression tests', () => {
    const wrapper = shallow(
      <Transition>
        <fakeChild />
      </Transition>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
