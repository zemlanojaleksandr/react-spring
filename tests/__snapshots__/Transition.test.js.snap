// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transition smoke/regression tests 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Transition
    keys={[Function]}
    reset={false}
    unique={false}
  >
    <fakeChild />
  </Transition>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "0",
    "nodeType": "function",
    "props": Object {
      "children": [Function],
      "config": undefined,
      "delay": 0,
      "filter": [Function],
      "from": undefined,
      "onFrame": undefined,
      "onRest": undefined,
      "onStart": undefined,
      "primitive": [Function],
      "reset": false,
      "state": "enter",
      "states": Object {
        "enter": undefined,
      },
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "Spring": [Function],
      "Trail": [Function],
      "create": [Function],
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "0",
      "nodeType": "function",
      "props": Object {
        "children": [Function],
        "config": undefined,
        "delay": 0,
        "filter": [Function],
        "from": undefined,
        "onFrame": undefined,
        "onRest": undefined,
        "onStart": undefined,
        "primitive": [Function],
        "reset": false,
        "state": "enter",
        "states": Object {
          "enter": undefined,
        },
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "Spring": [Function],
        "Trail": [Function],
        "create": [Function],
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
